### PR TITLE
Instance resolution made async.

### PR DESCRIPTION
**Pull request by SzymonPobiega**
The second stage of routing is about determining the list of instances given the endpoint name. The first stage has already been `async`-ed. Because there is possibility of users implementing that second stage based on some kind of remote resource (e.g. DB or web service lookup), it makes sense to make it `async` so that they can take advantage of async APIs of their storage technology.

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@janovesk @Particular/tf-async-core please review.

**Comment by danielmarbach** [comment]
Can you give a bit more context about the why?

On Thursday, December 10, 2015, Szymon Pobiega <notifications@github.com>
wrote:

> @janovesk <https://github.com/janovesk> @Particular/tf-async-core
> <https://github.com/orgs/Particular/teams/tf-async-core> please review.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3199#issuecomment-163707190>
> .
>


-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

**Comment by SzymonPobiega** [comment]
@danielmarbach updated the description.

File: src/NServiceBus.Core.Tests/Routing/SubscriptionRouterTests.cs [hunk 1]
```diff
@@ -31,18 +32,18 @@ public void Should_return_correct_base_and_inherited_address_even_if_they_differ
             publishers.AddStatic(inheritedEndpoint, baseType);
             publishers.AddStatic(inheritedEndpoint, inheritedType );
             var knownEndpoints = new EndpointInstances();
-            knownEndpoints.AddDynamic(e => new [] { new EndpointInstance(e, null, null) });
+            knownEndpoints.AddDynamic(e => Task.FromResult(EnumerableEx.Single(new EndpointInstance(e, null, null))));
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Routing/SubscriptionRouterTests.cs]: Why not just `knownEndpoints.AddDynamic(e => Task.FromResult(new [] { new EndpointInstance(e, null, null) }));`?

File: src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs [hunk 2]
```diff
@@ -43,7 +43,7 @@ protected internal override void Setup(FeatureConfigurationContext context)
 
                     var messageTypesHandled = GetMessageTypesHandledByThisEndpoint(handlerRegistry, conventions, settings);
 
-                    return new ApplySubscriptions(messageTypesHandled);
+                    return new ApplySubscriptions(messageTypesHandled, type => Task.FromResult(true));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs]: Not really related to this PR. But the AutoSubscribe feature is way to complex, envies way to many things it doesn't own and also reimplements some of the scanning logic. It feels very wrong. Can we do something here in another PR? @janovesk @SzymonPobiega

File: src/NServiceBus.Core/Routing/EndpointInstances.cs [hunk 3]
```diff
@@ -3,17 +3,23 @@ namespace NServiceBus.Routing
...
-        internal IEnumerable<EndpointInstance> FindInstances(Endpoint endpoint)
+        internal async Task<IEnumerable<EndpointInstance>> FindInstances(Endpoint endpoint)
         {
-            var distinctInstances = rules.SelectMany(r => r(endpoint)).Distinct().ToArray();
+            var instances = new List<EndpointInstance>();
+            foreach (var rule in rules)
+            {
+                instances.AddRange(await rule.Invoke(endpoint));
+            }
+            var distinctInstances = instances.Distinct().ToArray();
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/EndpointInstances.cs]: I think @SimonCropp raised this before. Do we need the distinct?

File: src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTableFeature.cs [hunk 4]
```diff
@@ -33,7 +33,7 @@ public StartupTask(ReadOnlySettings settings, FileRoutingTable routingTable)
 
             protected override Task OnStart(IBusContext contexts)
             {
-                settings.Get<EndpointInstances>().AddDynamic(e => routingTable.GetInstances(e));
+                settings.Get<EndpointInstances>().AddDynamic(e => Task.FromResult(routingTable.GetInstances(e)));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTableFeature.cs]: Would it be possible here to use `AddStatic`?

File: src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs [hunk 5]
```diff
@@ -80,21 +89,26 @@ static List<Type> GetMessageTypesHandledByThisEndpoint(MessageHandlerRegistry ha
...
             }
 
             protected override async Task OnStart(IBusContext context)
             {
-                foreach (var eventType in eventsToSubscribe)
+                foreach (var eventType in messagesHandledByThisEndpoint)
                 {
-                    await context.Subscribe(eventType).ConfigureAwait(false);
-                    Logger.DebugFormat("Auto subscribed to event {0}", eventType);
+                    if (await asyncPredicate(eventType))
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs]: `ConfigureAwait(false)`

File: src/NServiceBus.Core/Routing/EndpointInstances.cs [hunk 6]
```diff
@@ -3,17 +3,23 @@ namespace NServiceBus.Routing
...
+        List<Func<Endpoint, Task<IEnumerable<EndpointInstance>>>> rules = new List<Func<Endpoint, Task<IEnumerable<EndpointInstance>>>>();
 
-        internal IEnumerable<EndpointInstance> FindInstances(Endpoint endpoint)
+        internal async Task<IEnumerable<EndpointInstance>> FindInstances(Endpoint endpoint)
         {
-            var distinctInstances = rules.SelectMany(r => r(endpoint)).Distinct().ToArray();
+            var instances = new List<EndpointInstance>();
+            foreach (var rule in rules)
+            {
+                instances.AddRange(await rule.Invoke(endpoint));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/EndpointInstances.cs]: `ConfigureAwait(false)`

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscribeTerminator.cs [hunk 7]
```diff
@@ -23,11 +23,11 @@ public MessageDrivenSubscribeTerminator(SubscriptionRouter subscriptionRouter, s
             this.legacyMode = legacyMode;
         }
 
-        protected override Task Terminate(SubscribeContext context)
+        protected override async Task Terminate(SubscribeContext context)
         {
             var eventType = context.EventType;
 
-            var publisherAddresses = subscriptionRouter.GetAddressesForEventType(eventType)
+            var publisherAddresses = (await subscriptionRouter.GetAddressesForEventType(eventType))
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscribeTerminator.cs]: `ConfigureAwait(false)`

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscribeTerminator.cs [hunk 8]
```diff
@@ -52,8 +52,7 @@ protected override Task Terminate(SubscribeContext context)
 
                 subscribeTasks.Add(SendSubscribeMessageWithRetries(address, subscriptionMessage, eventType.AssemblyQualifiedName, context.Extensions));
             }
-
-            return Task.WhenAll(subscribeTasks.ToArray());
+            await Task.WhenAll(subscribeTasks.ToArray());
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscribeTerminator.cs]: `ConfigureAwait(false)`

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenUnsubscribeTerminator.cs [hunk 9]
```diff
@@ -23,11 +23,11 @@ public MessageDrivenUnsubscribeTerminator(SubscriptionRouter subscriptionRouter,
             this.legacyMode = legacyMode;
         }
 
-        protected override Task Terminate(UnsubscribeContext context)
+        protected override async Task Terminate(UnsubscribeContext context)
         {
             var eventType = context.EventType;
 
-            var publisherAddresses = subscriptionRouter.GetAddressesForEventType(eventType)
+            var publisherAddresses = (await subscriptionRouter.GetAddressesForEventType(eventType))
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenUnsubscribeTerminator.cs]: `ConfigureAwait(false)`

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenUnsubscribeTerminator.cs [hunk 10]
```diff
@@ -50,8 +50,7 @@ protected override Task Terminate(UnsubscribeContext context)
 
                 unsubscribeTasks.Add(SendUnsubscribeMessageWithRetries(publisherAddress, unsubscribeMessage, eventType.AssemblyQualifiedName, context.Extensions));
             }
-
-            return Task.WhenAll(unsubscribeTasks.ToArray());
+            await Task.WhenAll(unsubscribeTasks.ToArray());
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenUnsubscribeTerminator.cs]: `ConfigureAwait(false)`

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/PublisherAddress.cs [hunk 11]
```diff
@@ -58,7 +59,8 @@ internal IEnumerable<string> Resolve(Func<Endpoint, IEnumerable<EndpointInstance
             {
                 return instances.Select(addressResolver);
             }
-            return instanceResolver(endpoint).Select(addressResolver);
+            var result = await instanceResolver(endpoint);
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/PublisherAddress.cs]: `ConfigureAwait(false)`
Is there no way we can turn this into a return without await?

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/SubscriptionRouter.cs [hunk 12]
```diff
@@ -14,11 +14,16 @@ public SubscriptionRouter(Publishers publishers, EndpointInstances endpointInsta
...
+        public async Task<IEnumerable<string>> GetAddressesForEventType(Type messageType)
         {
-            return publishers
-                .GetPublisherFor(messageType).SelectMany(p => p
-                    .Resolve(e => endpointInstances.FindInstances(e), i => physicalAddresses.GetTransportAddress(i)));
+            var results = new List<string>();
+            foreach (var publisherAddress in publishers.GetPublisherFor(messageType))
+            {
+                results.AddRange(await publisherAddress.Resolve(
+                    async e => await endpointInstances.FindInstances(e).ConfigureAwait(false),
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/SubscriptionRouter.cs]: Convert to Method Group and then you probably don't need the await

Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/SubscriptionRouter.cs]: But still `ConfigureAwait(false)` missing

File: src/NServiceBus.Core/Routing/UnicastRoute.cs [hunk 13]
```diff
@@ -42,24 +44,18 @@ public UnicastRoute(string physicalAddress)
...
-            else
-            {
-                var instances = instanceResolver(endpoint);
-                foreach (var i in instances)
-                {
-                    yield return UnicastRoutingTarget.ToEndpointInstance(i);
-                }
+                return EnumerableEx.Single(UnicastRoutingTarget.ToEndpointInstance(instance));
             }
+            var instances = await instanceResolver(endpoint);
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/UnicastRoute.cs]: `ConfigureAwait(false)`
Is there no way we can turn this into a return without await?

File: src/NServiceBus.Core/Routing/UnicastRouter.cs [hunk 14]
```diff
@@ -31,10 +31,13 @@ public UnicastRouter(MessageMetadataRegistry messageMetadataRegistry, UnicastRou
                 .ToList();
             
             var routes = await unicastRoutingTable.GetDestinationsFor(typesToRoute, contextBag).ConfigureAwait(false);
+            var destinations = new List<UnicastRoutingTarget>();
+            foreach (var route in routes)
+            {
+                destinations.AddRange(await route.Resolve(async e => await endpointInstances.FindInstances(e)));
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/UnicastRouter.cs]: Convert to Method Group and then you probably don't need the await

Comment (danielmarbach) [inline src/NServiceBus.Core/Routing/UnicastRouter.cs]: But still `ConfigureAwait(false)` missing

**Comment by danielmarbach** [comment]
@SzymonPobiega You owe me 10 (missing ConfigureAwait(false)) umbrella drinks in Thailand ;)

**Comment by andreasohlund** [comment]
The :beer: count is getting worrying for @SzymonPobiega even given the Thailand prices ;)

File: src/NServiceBus.Core.Tests/Routing/SubscriptionRouterTests.cs (hunk 1 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core.Tests/Routing/SubscriptionRouterTests.cs]: Becate that would require casting to `IEnumerable<T>` which is sooo long.

File: src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs (hunk 2 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs]: @danielmarbach can you raise an issue in the NSB repo for that? I agree it is wrong.

File: src/NServiceBus.Core/Routing/EndpointInstances.cs (hunk 3 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/EndpointInstances.cs]: Here we do need it because various sources can provide the same instance and downstream code expects the items to be unique.

File: src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTableFeature.cs (hunk 4 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTableFeature.cs]: Nope, because the file is monitored and when the contents change, we want that change be reflected asap, not after the restart of the endpoint. Currently there is no way to remove a statically added instance.

File: src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs (hunk 5 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs]: :+1:

File: src/NServiceBus.Core/Routing/EndpointInstances.cs (hunk 6 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/EndpointInstances.cs]: :+1:

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscribeTerminator.cs (hunk 7 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscribeTerminator.cs]: :+1:

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscribeTerminator.cs (hunk 8 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscribeTerminator.cs]: :+1:

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenUnsubscribeTerminator.cs (hunk 9 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenUnsubscribeTerminator.cs]: :+1:

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenUnsubscribeTerminator.cs (hunk 10 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenUnsubscribeTerminator.cs]: :+1:

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/PublisherAddress.cs (hunk 11 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/PublisherAddress.cs]: Fixed `ConfigureAwait(false)`. Not sure how I can get rid of `await'

File: src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/SubscriptionRouter.cs (hunk 12 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/SubscriptionRouter.cs]: I think I fixed it.

File: src/NServiceBus.Core/Routing/UnicastRoute.cs (hunk 13 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/UnicastRoute.cs]: Added `await`. Not sure how to convert :(

File: src/NServiceBus.Core/Routing/UnicastRouter.cs (hunk 14 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Routing/UnicastRouter.cs]: I think I fixed it.

**Comment by SzymonPobiega** [comment]
@danielmarbach I think I implemented all your great comments :-)

File: src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs [hunk 15]
```diff
@@ -17,32 +18,32 @@ public void Should_throw_when_trying_to_configure_instances_that_do_not_match_en
...
         [Test]
-        public void Should_filter_out_duplicate_instances()
+        public async Task Should_filter_out_duplicate_instances()
         {
             var instances = new EndpointInstances();
             var sales = new EndpointName("Sales");
             instances.AddStatic(sales, new EndpointInstance(sales, "dup", null), new EndpointInstance(sales, "dup"));
 
-            var salesInstances = instances.FindInstances(sales).ToList();
+            var salesInstances = (await instances.FindInstances(sales)).ToList();
```
Comment (SimonCropp) [inline src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs]: To list is redundant. Just use on the enumerable

File: src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs [hunk 16]
```diff
@@ -17,32 +18,32 @@ public void Should_throw_when_trying_to_configure_instances_that_do_not_match_en
...
         [Test]
-        public void Should_return_instances_configured_by_static_route()
+        public async Task Should_return_instances_configured_by_static_route()
         {
             var instances = new EndpointInstances();
             var sales = new EndpointName("Sales");
             instances.AddStatic(sales, new EndpointInstance(sales, "1", null), new EndpointInstance(sales, "2"));
 
-            var salesInstances = instances.FindInstances(sales).ToList();
+            var salesInstances = (await instances.FindInstances(sales)).ToList();
```
Comment (SimonCropp) [inline src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs]: To list is redundant. Just use on the enumerable

File: src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs [hunk 17]
```diff
@@ -17,32 +18,32 @@ public void Should_throw_when_trying_to_configure_instances_that_do_not_match_en
...
             Assert.AreEqual(1, salesInstances.Count);
         }
 
         [Test]
-        public void Should_default_to_single_instance_when_not_configured()
+        public async Task Should_default_to_single_instance_when_not_configured()
         {
             var instances = new EndpointInstances();
-            var salesInstances = instances.FindInstances(new EndpointName("Sales")).ToArray();
+            var salesInstances = (await instances.FindInstances(new EndpointName("Sales"))).ToArray();
```
Comment (SimonCropp) [inline src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs]: Just do a single. It will verify a single instance and simplify the code

File: src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs (hunk 16 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs]: :+1:

File: src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs (hunk 15 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs]: :+1:

File: src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs (hunk 17 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs]: :+1:

**Comment by SzymonPobiega** [comment]
@SimonCropp done. Rebased. Feel free to push the green button.
